cli/src/action/login/oauth/github.rs

```diff
@@ -0,0 +1,238 @@
+
+pub struct GithubProvider {
+    client_id: Option<&'static str>,
+    client_secret: Option<&'static str>,
+}
+
+impl OAuth2Provider for GithubProvider {
+    fn get_tokens(&self) -> Result<UserTokens, OAuth2ProviderError> {
+        let client_id = self.client_id.ok_or_else(|| {
+            OAuth2ProviderError("Github client id was not configured for this binary".into())
+        })?;
+        let client_secret = self.client_secret.ok_or_else(|| {
+            OAuth2ProviderError("Github client secret was not configured for this binary".into())
+        })?;
+
+        let callback = callback::OAuth2Callback::new().expect("Unable to start callback");
+
+        let auth_url = AuthUrl::new("https://github.com/login/oauth/authorize".to_string())
+            .map_err(|err| OAuth2ProviderError(err.to_string()))?;
+        let token_url = TokenUrl::new("https://github.com/login/oauth/access_token".to_string())
+            .map_err(|err| OAuth2ProviderError(err.to_string()))?;
+        let redirect_url = RedirectUrl::new(format!("http://{}/auth_response", callback.address()))
+            .map_err(|err| OAuth2ProviderError(err.to_string()))?;
+
+        let client = BasicClient::new(
+            ClientId::new(client_id.to_string()),
+            Some(ClientSecret::new(client_secret.to_string())),
+            auth_url,
+            Some(token_url),
+        )
+        .set_redirect_url(redirect_url);
+
+        // Github supports PKCE challenge.
+        let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+        // Generate the full authorization URL.
+        let (auth_url, _csrf_token) = client
+            .authorize_url(CsrfToken::new_random)
+            // Set the desired scopes.
+            .add_scope(Scope::new("user:email".to_string()))
+            // Set the PKCE code challenge.
+            .set_pkce_challenge(pkce_challenge)
+            .url();
+
+        info!("Opening {}...", auth_url);
+        webbrowser::open(&auth_url.to_string()).map_err(|err| {
+            debug!("Unable to open authorization URL: {}", err);
+            OAuth2ProviderError(
+                "Unable to open Authorization URL in browser; \
+                ensure that a default browser is installed."
+                    .into(),
+            )
+        })?;
+
+        let auth_code = callback
+            .receive_authorization_code()
+            .expect("Unable receive authorization code");
+
+        let token_result = client
+            .exchange_code(AuthorizationCode::new(auth_code))
+            .set_pkce_verifier(pkce_verifier)
+            .request(|req| {
+                Client::new()
+                    .execute(to_reqwest_req(req)?)
+                    .map(to_oauth2_res)
+                    .map_err(|_err| {
+                        OAuth2ProviderError("Unable to execute for OAuth2 Access Token".into())
+                    })
+            })
```

Comment:
You can do `.request(oauth2::reqwest::http_client)` to simplify this

Reply:
The default features are disabled, and I made this choice since we already had reqwest as a dependency.  Did not seem necessary to pull in multiple versions of reqwest (or use the alternative of a curl client).